### PR TITLE
💥 Implement hooks

### PR DESCRIPTION
**Pull request by randycoulman**
### Describe the purpose of your pull request

Implements the new hooks feature to allow users to supply callbacks that are called when significant events occur.

I based this implementation largely off of the Ruby SDK with necessary adjustments to accommodate Elixir's functional nature.

In particular, I chose to implement `ConfigCat.hooks`, which returns an opaque value (actually a `ConfigCat.instance_id`) that can later be passed to `Hooks.add_*_callback` as this is consistent with the other SDKs. I think a better option for Elixir would be to add the `add_*_callback` functions directly to `ConfigCat` instead. That way, `ConfigCat` can continue to serve as a facade for the entire library and users don't have to worry about calling functions on other modules.  Let me know if you'd prefer this option.

In order to store the configured hooks separately for each instance, I made `Hooks` into a GenServer, but ensured that the hooks are invoked in the calling process rather than in the GenServer's process. That way, the GenServer doesn't become a bottleneck.

The `on_error` callback is not yet called everywhere it should be. In order to call it, we need an `instance_id` and it's not easy to get it everywhere it needs to be.  I have an idea to address this that will also address #51, so I'll do that in a follow-up PR.

### Related issues (only if applicable)

https://trello.com/c/FRaC8jzd/7-hooks-elixir

### Requirement checklist (only if applicable)

- [x] I have covered the applied changes with automated tests.
- [x] I have executed the full automated test set against my changes.
- [x] I have validated my changes against all supported platform versions.
- [x] I have read and accepted the [contribution agreement](https://github.com/configcat/legal/blob/main/contribution-agreement.md).

### CODE REVIEW

File: test/config_cat/cache_policy/auto_test.exs [hunk 1]
```diff
@@ -92,51 +82,6 @@ defmodule ConfigCat.CachePolicy.AutoTest do
...
-
-      {:ok, _} = start_cache_policy(policy)
-
-      assert_receive(:callback)
-
-      expect_unchanged()
-      wait_for_poll(policy)
-
-      refute_receive(:callback)
-    end
```
Comment (randycoulman) [inline test/config_cat/cache_policy/auto_test.exs:139]: These two tests were moved down and modified to use hooks.

File: test/config_cat/cache_policy/auto_test.exs [hunk 2]
```diff
@@ -153,56 +98,74 @@ defmodule ConfigCat.CachePolicy.AutoTest do
       assert {:ok, settings, entry.fetch_time_ms} == CachePolicy.get(instance_id)
     end
 
-    test "calls the change callback", %{entry: entry} do
```
Comment (randycoulman) [inline test/config_cat/cache_policy/auto_test.exs:156]: This test was also moved down.

File: test/config_cat/cache_policy/auto_test.exs [hunk 3]
```diff
@@ -153,56 +98,74 @@ defmodule ConfigCat.CachePolicy.AutoTest do
...
-      :ok = CachePolicy.force_refresh(instance_id)
+      {:ok, instance_id} = start_cache_policy(@policy)
 
-      assert_receive(:callback)
+      assert_returns_error(fn -> CachePolicy.force_refresh(instance_id) end)
     end
+  end
 
-    @tag capture_log: true
-    test "handles errors in the change callback", %{entry: entry, settings: settings} do
```
Comment (randycoulman) [inline test/config_cat/cache_policy/auto_test.exs:173]: This test was removed; the exception-handling behavior of hooks is well-tested elsewhere.

File: test/config_cat/hooks_test.exs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    _hooks =
+      ConfigCat.hooks(client: instance_id)
+      |> Hooks.add_on_client_ready({TestHooks, :on_client_ready, [test_pid]})
+      |> Hooks.add_on_config_changed({TestHooks, :on_config_changed, [test_pid]})
+      |> Hooks.add_on_flag_evaluated({TestHooks, :on_flag_evaluated, [test_pid]})
+      |> Hooks.add_on_error({TestHooks, :on_error, [test_pid]})
+
+    :ok = start_client(instance_id: instance_id)
+
+    value = ConfigCat.get_value("testStringKey", "", client: instance_id)
```
Comment (randycoulman) [inline test/config_cat/hooks_test.exs:112]: This line occasionally fails in local development. There seems to be some kind of race condition that happens that I haven't yet figured out. I'll keep an eye on it and see what I can learn.

File: lib/config_cat/client.ex [hunk 5]
```diff
@@ -202,31 +203,35 @@ defmodule ConfigCat.Client do
...
+
+        _ ->
+          message =
+            "Config JSON is not present when evaluating setting '#{key}'. Returning the `default_value` parameter that you specified in your application: '#{default_value}'."
+
+          EvaluationDetails.new(
+            default_value?: true,
+            error: message,
+            key: key,
+            value: default_value,
```
Comment (kp-cat) [inline lib/config_cat/client.ex:229]: Could you please add user in error case too. It seems this is a bug in the Ruby SDK I'll fix it there too.
```suggestion
            user: user,
            value: default_value,
```

Comment (kp-cat) [reply lib/config_cat/client.ex:229]: Could you please add the user to the details object in the `def evaluate(key, user, default_value, default_variation_id, settings)` too? Unfortunately, I can't see this file in this PR and can't give a suggestion for that.

Comment (randycoulman) [reply lib/config_cat/client.ex:229]: Done in 485f346

**Review by kp-cat — APPROVED** [review]

**Pull request merged**